File: C/0021-merge-two-sorted-lists.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* mergeTwoLists(struct ListNode *list1, struct ListNode *list2) {
    struct ListNode *result;
    struct ListNode *last;

    if(list1 == NULL) {
        return list2;
    }  
    if(list2 == NULL) {
        return list1;
    }

    if(list1 -> val <= list2 -> val) {
        result = list1;
        list1 = list1 -> next;
    } else {
        result = list2;
        list2 = list2 -> next;
    }

    last = result;

    while(list1 != NULL && list2 != NULL) {
        if(list1 -> val <= list2 -> val) {
            last -> next = list1;
            list1 = list1 -> next;
        } else {
            last -> next = list2;
            list2 = list2 -> next;
        }
        last = last -> next;
    }

    if(list1 != NULL) {
        last -> next = list1;
    }

    if(list2 != NULL) {
        last -> next = list2;
    }

    return result;
}

File: dev-rust/0021-merge-two-sorted-lists.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn merge_two_lists(list1: Option<Box<ListNode>>, list2: Option<Box<ListNode>>) -> Option<Box<ListNode>> { return match (list1, list2) {
        (None, None) => {
            return None;
        }

        (Some(l), None) => {
            return Some(l);
        }

        (None, Some(l)) => {
            return Some(l);
        }

        (Some(l1), Some(l2)) => {
            if l1.val <= l2.val {
                let tail = Self::merge_two_lists(l1.next, Some(l2));
                return Some(Box::new(ListNode { next: tail, val: l1.val }));
            } else {
                let tail = Self::merge_two_lists(Some(l1), l2.next);
                return Some(Box::new(ListNode { next: tail, val: l2.val }));
            }

            return None;
        }

    }}
}
